### ISSUE TITLE
Create Pre-production environment for Cloud

### ISSUE THREAD

**Issue by daxmc99** [sourcegraph/sourcegraph#28082]
Links https://sourcegraph.atlassian.net/browse/CLOUD-64

## TODO:
- [x] Terraform infrastructure in `sourcegraph-pre-prod` GCP project
- [x] [Create overlays](https://github.com/sourcegraph/sourcegraph/issues/29255)
- [x] Upgrade ingress resource https://github.com/sourcegraph/sourcegraph/issues/29315
- [x] Create DNS name for `pre-prod` 
- [x] Duplicate k8s.sgdev.org database (no customer data) 
- [x] [Replicate secrets](https://github.com/sourcegraph/sourcegraph/issues/29981)
- [x] Deploy Sourcegraph into this cluster
- [x] Develop steps to promote from pre-prod

~~Blocked by: https://github.com/sourcegraph/sourcegraph/issues/27036~~

**Comment by danieldides** [comment]
Should we add a the `tracking` label to this issue and start making sub-issues out of the stuff we came up with Friday morning?

**Comment by filiphaftek** [comment]
**CloudSQL** terraform module - I review existing one and looks very good [module link](https://github.com/sourcegraph/infrastructure/tree/main/modules/sql).
It has:
- HA setup
- resizeable disk
- private connectivity for GKE nodes
- public ip4 endpoint, which allows to connect from the outside (using gcloud or sqlproxy)
- it also exposes parameter for IP ranges whitelisting, for now not used (anyone can connect if has access to GCP project)

Some things missing:
- sql user creation - [docs](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/sql_database_instance)
```
NOTE on google_sql_database_instance:
- Second-generation instances include a default 'root'@'%' user with no password. This user will be deleted by Terraform on instance creation. You should use google_sql_user to define a custom user with a restricted host and strong password.
```
so the only thing is that all environments has to create separate `google_sql_user` with password explicitly and add it to the sql instance.
I will make separate PR just for creating database in pre-prod using this module - @daxmc99 do we know already the expected size of pre-prod? in terms of SQL instance size?

**Comment by daxmc99** [comment]
@filiphaftek I believe the disks are resizable. I would default to 100 GB and let it size up from there. We are going to use internal databases that are smaller than prod. 
Current size is: `2.8G` from `kubectl --context=dogfood exec pgsql-6f54c7f676-lwbgs --it -- df -h`

**Comment by daxmc99** [comment]
Flow for pre-prod: 
1. Separate branch "pre-prod" which has the newest images. 
2. Create automation to merge the pre-prod branch to release - ensure that the change is only an image change. 
3. Validate the changes by dumping the entire kustomize YAML build (should allow for easier to view diffs, expect to see mostly image only updates) 

Still need to fine tune the promotion step to pre-prod.

**Comment by rafax** [comment]
@daxmc99 will we switch existing CI to build&deploy to pre-prod instead of prod? 
How / how often will we promote from pre-prod to prod?

**Comment by daxmc99** [comment]
Our pre-prod promotion strategy:

1. Long-lived `preprod` branch where we update images as quickly as possible. 
2. `release` branch where we merge from preprod on a schedule
2a. Merges are peformed automatically if and only if the changes are _only_ image updates otherwise approval is needed

This can be done entirely in our CI tooling today I believe

**Comment by daxmc99** [comment]
Miro board: https://miro.com/app/board/uXjVOKqPSDM=/ 

This work tracks 2 items to achieve SOC 2 compliance: 
- Automated Pull requests to `pre-prod` branch (Renovate)
- Automated promotion to prod (cronjob / custom event)

**Comment by jhchabran** [comment]
@danieldides / @daxmc99 is there a ticket tracking the work on `sg ops` so we can move to continuously deploy the preprod?

**Comment by filiphaftek** [comment]
@jhchabran we have an [issue](https://github.com/sourcegraph/sourcegraph/issues/29251) for it, but as we have Renovate + GithuActions working fully automated now, maybe we could close this stage of preprod and open new tracking issue for `preprod` improvements? This could contains `sg ops`, automated rollback etc. 
@danieldides @daxmc99  what do you think?
